### PR TITLE
Change hook stderr output to be logged at WARNING level

### PR DESCRIPTION
**Pull request by jameinel**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [ ] Do comments answer the question of why design decisions were made?

----

## Description of change

We have been logging both stdout and stderr for units at DEBUG, but that
leaves everything at debug. And while we want to stop including DEBUG by default, we do want to include things like Tracebacks by default. However those are written to stderr.
So we can split the logging, and log all of stdout to DEBUG and stderr to
WARNING.
This also fixes the default logging level, which had been "logger.LoggerInfo()"
which defaults to `<root>=WARNING` and sets it to what we intended it to be
`<root>=INFO`.

## QA steps

```sh
$ juju bootstrap lxd lxd
$ juju model-config logging-config
<root>=INFO
$ juju model-config -m controller logging-config
<root>=INFO
# Edit a charm to inject a python exception
$ juju deploy broken-charm
$ juju debug-log --include broken-charm/0 --replay
```

That should show lines something like:
```unit-uo-0: 01:12:30 INFO juju.worker.uniter awaiting error resolution for "install" hook
unit-uo-0: 01:13:54 INFO juju.worker.uniter awaiting error resolution for "install" hook
unit-uo-0: 01:13:54 WARNING unit.uo/0.install Traceback (most recent call last):
unit-uo-0: 01:13:54 WARNING unit.uo/0.install   File "/var/lib/juju/agents/unit-uo-0/charm/dispatch", line 14,in <module>
unit-uo-0: 01:13:54 WARNING unit.uo/0.install     import broken
unit-uo-0: 01:13:54 WARNING unit.uo/0.install ModuleNotFoundError: No module named 'broken'
unit-uo-0: 01:13:54 ERROR juju.worker.uniter.operation hook "install" (via hook dispatching script: dispatch) f
```

## Documentation changes

We should note the change in behavior for logging.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1893993

### CODE REVIEW

**Comment by jameinel** [comment]
With my patch and a charm whose src/charm.py has an intentional bug in imports I get:
```
unit-uo-0: 23:02:20 INFO juju.worker.uniter found queued "install" hook
unit-uo-0: 23:02:20 WARNING unit.uo/0.install Traceback (most recent call last):
unit-uo-0: 23:02:20 WARNING unit.uo/0.install   File "/var/lib/juju/agents/unit-uo-0/charm/dispatch", line 14, in <module>
unit-uo-0: 23:02:20 WARNING unit.uo/0.install     import broken
unit-uo-0: 23:02:20 WARNING unit.uo/0.install ModuleNotFoundError: No module named 'broken'
unit-uo-0: 23:02:20 ERROR juju.worker.uniter.operation hook "install" (via hook dispatching script: dispatch) failed: exit status 1
```

**Review by hpidcock — APPROVED** [review]
LGTM. One thing I'm not certain on this change is how it affects action output. Action output should be pristine as it could be consumed by scripts using the CLI. Could you confirm this?

File: environs/config/config.go [hunk 1]
```diff
@@ -516,18 +516,15 @@ func ConfigDefaults() map[string]interface{} {
...
+func (c *Config) setLoggingFromEnviron() error {
 	loggingConfig := c.asString("logging-config")
 	// If the logging config hasn't been set, then look for the os environment
 	// variable, and failing that, get the config from loggo itself.
 	if loggingConfig == "" {
 		if environmentValue := os.Getenv(osenv.JujuLoggingConfigEnvKey); environmentValue != "" {
-			loggingConfig = environmentValue
-		} else {
-			loggingConfig = loggo.LoggerInfo()
+			c.defined["logging-config"] = loggingConfig
```
Comment (wallyworld) [inline environs/config/config.go:525]: s/loggingConfig/environmentValue

Comment (wallyworld) [reply environs/config/config.go:525]: Seems like there should be a failing test for this typo, it not maybe we need one to check that config from environ works

Comment (wallyworld) [reply environs/config/config.go:525]: Ah, so in the TestLoggingConfigFromEnvironment test, we do

s.PatchEnvironment(osenv.JujuLoggingConfigEnvKey, "<root>=INFO")

this is now the default value so we need to fix the test to patch something else.

File: worker/common/charmrunner/logger.go [hunk 2]
```diff
@@ -63,6 +63,12 @@ func (l *HookLogger) Run() {
 	}
 }
 
+func (l *HookLogger) AddReceiver(receiver MessageReceiver) {
```
Comment (wallyworld) [inline worker/common/charmrunner/logger.go:66]: comment

File: worker/uniter/runner/runner.go [hunk 3]
```diff
@@ -511,6 +511,12 @@ func (b *bufferAdaptor) Messagef(isPrefix bool, message string, args ...interfac
 	b.outCopy.WriteString(formattedMessage)
 }
 
+func (b *bufferAdaptor) Bytes() []byte {
```
Comment (wallyworld) [inline worker/uniter/runner/runner.go:514]: ideally comment public funcs but i see we're not consistent with that in this file

File: worker/uniter/runner/runner.go [hunk 4]
```diff
@@ -656,28 +658,10 @@ func (runner *runner) runCharmProcessOnLocal(hook, hookName, charmDir string, en
 
 	// If we are running an action, record stdout and stderr.
 	if runningAction {
-		readBytes := func(r io.Reader) []byte {
```
Comment (wallyworld) [inline worker/uniter/runner/runner.go:659]: thanks

**Review by wallyworld — APPROVED** [review]
Just a small fix needed

File: worker/uniter/runner/runner.go (hunk 3 above)
Comment (SimonRichardson) [reply worker/uniter/runner/runner.go:514]: But that doesn't mean we shouldn't in the future :wink:

**Comment by jameinel** [comment]
!!build!!

**Comment by jameinel** [comment]
!!build!!

gophertest failed to build?
```
10:39:07 exit status 1
10:39:07 github.com/hpidcock/gophertest/cache/puller.(*Puller).Visit
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/cache/puller/puller.go:77
10:39:07 github.com/hpidcock/gophertest/dag.(*DAG).VisitAll.func1
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/dag/dag.go:468
10:39:07 golang.org/x/sync/errgroup.(*Group).Go.func1
10:39:07 	/home/ubuntu/go/src/golang.org/x/sync/errgroup/errgroup.go:57
10:39:07 runtime.goexit
10:39:07 	/usr/local/go1.14.4/src/runtime/asm_amd64.s:1373
10:39:07 github.com/hpidcock/gophertest/dag.(*DAG).VisitAll
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/dag/dag.go:474
10:39:07 main.Main
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/main.go:221
10:39:07 main.main
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/main.go:56
10:39:07 runtime.main
10:39:07 	/usr/local/go1.14.4/src/runtime/proc.go:203
10:39:07 runtime.goexit
10:39:07 	/usr/local/go1.14.4/src/runtime/asm_amd64.s:1373
10:39:07 pulling from cache
10:39:07 main.Main
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/main.go:223
10:39:07 main.main
10:39:07 	/home/ubuntu/go/src/github.com/hpidcock/gophertest/main.go:56
10:39:07 runtime.main
10:39:07 	/usr/local/go1.14.4/src/runtime/proc.go:203
10:39:07 runtime.goexit
10:39:07 	/usr/local/go1.14.4/src/runtime/asm_amd64.s:1373++ date --rfc-3339=seconds
10:39:07 + echo 2020-09-10 15:39:07+00:00 'running gopher.test'
10:39:07 2020-09-10 15:39:07+00:00 running gopher.test
10:39:07 +++ nproc
10:39:07 ++ expr 16 / 2
10:39:07 + ./gopher.test -c 8 -- -check.v
10:39:07 bash: line 37: ./gopher.test: No such file or directory
10:39:07 + cleanup
```

**Comment by jameinel** [comment]
$$merge$$

**Pull request merged**